spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/ReplaceData.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.catalyst.analysis.NamedRelation
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.util.CharVarcharUtils
+import org.apache.spark.sql.connector.write.Write
+import org.apache.spark.sql.types.DataType
+
+/**
+ * Replace data in an existing table.
+ */
+case class ReplaceData(
+    table: NamedRelation,
+    query: LogicalPlan,
+    originalTable: NamedRelation,
+    write: Option[Write] = None) extends V2WriteCommandLike {
+
+  override lazy val resolved: Boolean = table.resolved && query.resolved && outputResolved
+  override lazy val stringArgs: Iterator[Any] = Iterator(table, query, write)
+
+  // the incoming query may include metadata columns
+  lazy val dataInput: Seq[Attribute] = {
```

Comment:
Can you give me a little more context on this?

Reply:
This node has custom resolution as the input query may have more columns than the table. For example, below the query contains `id#249, dep#250, _file#254, _pos#255L`.

```
+- ReplaceData RelationV2[id#249, dep#250] testhive.default.table
   +- Filter NOT ((id#249 IN (list#248 []) AND (dep#250 = hr)) <=> true)
      :  +- Project [(value#19 + 2) AS (value + 2)#251]
      :     +- SubqueryAlias deleted_id
      :        +- View (`deleted_id`, [value#19])
      :           +- LocalRelation [value#19]
      +- RelationV2[id#249, dep#250, _file#254, _pos#255L] testhive.default.table
```

I thought I was projecting away the metadata columns in the writer but I am no longer sure. I'll have to check.

Reply:
Yeah, I do have a projection in `ExtendedV2Writes`.

Reply:
If there is a projection before the replace data, then we don't need to modify this?

Reply:
The projection is inserted only after the write is built as the extra columns can be part of the distribution and ordering requested by the write.

Reply:
In other words, we can discard the metadata columns only after the write has been built in the optimizer. That's why I added a projection in `ExtendedV2Writes`. Since the rewrite happens in the analyzer,  `ReplaceData` must be considered resolved as long as the data columns are aligned. That's why this node has custom resolution.